modules/standard/IO.chpl

```diff
@@ -3615,65 +3620,207 @@ private inline proc _write_one_internal(_channel_internal:qio_channel_ptr_t, par
+  // Create a new channel that borrows the pointer in the
+  // existing channel so we can avoid locking (because we
+  // already have the lock)
+  var writer = new channel(writing=true, iokind.dynamic, locking=false,
+                           home=here,
+                           _channel_internal=_channel_internal);
+
   var err:syserr = ENOERR;
-  var save_style:iostyle;
-  var saved_style = false;
 
-  // Adjust the style to use Chapel strings in quotes
-  // unless the string style was already overridden
-  // (that way, writef("%j") will still use JSON-style strings)
-  if qio_channel_str_style(_channel_internal) == QIO_STRING_FORMAT_WORD {
-    qio_channel_get_style(_channel_internal, save_style);
-    var mystyle = save_style;
-    mystyle.string_format = QIO_STRING_FORMAT_CHPL;
-    qio_channel_set_style(_channel_internal, mystyle);
-    saved_style = true;
+  // Clear the channel error so we can use the error
+  // to stop writing if there was an error.
+  qio_channel_clear_error(_channel_internal);
+
+  if (isClassType(t) || chpl_isDdata(t)) && x == nil {
+    // future - write class IDs, have serialization format
+    var st = writer.styleElement(QIO_STYLE_ELEMENT_AGGREGATE);
+    var iolit:ioLiteral;
+    if st == QIO_AGGREGATE_FORMAT_JSON {
+      iolit = new ioLiteral("null");
+    } else {
+      iolit = new ioLiteral("nil");
+    }
+    _write_one_internal(_channel_internal, iokind.dynamic, iolit);
+  } else {
+    x.writeThis(writer);
   }
 
-  writer.write(x);
+  // Set the channel pointer to NULL to make the
+  // destruction of the local writer record safe
+  // (it shouldn't release anything since it's a local copy).
+  writer._channel_internal = QIO_CHANNEL_PTR_NULL;
+
+  err = qio_channel_error(_channel_internal);
+  return err;
+}
+
+pragma "no doc"
+proc channel.readIt(ref x) {
+  if writing then compilerError("read on write-only channel");
+  on this.home {
+    this.lock();
+    var error:syserr;
+    error = qio_channel_error(_channel_internal);
+    if ! error {
+      error = _read_one_internal(_channel_internal, kind, x);
+      _qio_channel_set_error_unlocked(_channel_internal, error);
+    }
+    this.unlock();
+  }
+}
 
-  // Put the style back
-  if saved_style {
-    qio_channel_set_style(_channel_internal, save_style);
+pragma "no doc"
+proc channel.writeIt(x) {
+  if !writing then compilerError("write on read-only channel");
+  on this.home {
+    this.lock();
+    var error:syserr;
+    error = qio_channel_error(_channel_internal);
+    if ! error {
+      error = _write_one_internal(_channel_internal, kind, x);
+      _qio_channel_set_error_unlocked(_channel_internal, error);
+    }
+    this.unlock();
   }
+}
 
-  err = writer.err;
-  delete writer;
 
-  return err;
+inline proc channel.readwrite(x) where this.writing {
+  this.writeIt(x);
+}
+inline proc channel.readwrite(ref x) where !this.writing {
+  this.readIt(x);
 }
 
+  inline proc <~>(w: channel, x) where w.writing {
```

Comment:
It looks like you lost the documentation comments for these functions when you remade them.